includes/express-checkout/class-wc-payments-express-checkout-button-handler.php

```diff
@@ -400,10 +400,9 @@ public function display_pay_for_order_page_html( $order ) {
 	 * @return void
 	 */
 	public function set_session() {
-		// Don't set session cookies on product pages to allow for caching when express checkout
-		// buttons are disabled. But keep cookies if there is already an active WC session in place.
+		// Only set session cookies on checkout and cart pages, but skip if there's already an active WC session.
 		if (
-			! ( $this->express_checkout_helper->is_product() && $this->express_checkout_helper->should_show_express_checkout_button() )
+			! ( $this->express_checkout_helper->is_checkout() || $this->express_checkout_helper->is_cart() )
 			|| ( isset( WC()->session ) && WC()->session->has_session() )
```

Comment:
I think it would be easier to read if existing WooCommerce session check will come first.
```suggestion
			( isset( WC()->session ) && WC()->session->has_session() )
			|| ! ( $this->express_checkout_helper->is_checkout() || $this->express_checkout_helper->is_cart() )
```

Maybe even create a variable `$should_render_express_checkout_button = $this->express_checkout_helper->is_checkout() || $this->express_checkout_helper->is_cart()` and use it in the check, to make meaning clear.

Reply:
Good point! I clarified this in [commit a07d0d8](https://github.com/Automattic/woocommerce-payments/pull/10747/commits/a07d0d86cd413af2fa71393a3035513d537de715).
I'm not using the `$should_render_express_checkout_button suggestion` because it might be misleading—express checkout buttons are also rendered on product pages, not just in the checkout.
Let me know what you think!